lib/puppet/provider.rb

```diff
@@ -292,13 +292,25 @@ def self.default_match
   end
 
   def self.fact_match(fact, values)
+
     values = [values] unless values.is_a? Array
-    values.map! { |v| v.to_s.downcase.intern }
+
+    values.map! do |v|
+      if v.is_a? Regexp
+        v
+      else
+        v.to_s.downcase.intern
+      end
+    end
 
     if fval = Facter.value(fact).to_s and fval != ""
       fval = fval.to_s.downcase.intern
-
-      values.include?(fval)
+      if values.any? {|v| v.is_a? Regexp }
+        regex_match = Regexp.union(values)
+        fval =~ regex_match
```

Comment:
So, I ended up asking a question over on Code Review on Stack Exchange, because I felt this logic was overly complex, any I got a suggestion that would fix the above issue _and_ make he code simpler:

```
fact_val = Facter.value(fact).to_s.downcase
fact_val != "" and [*values].any? { |v| v === fact_val }
```

So this works, and all the unit tests pass, except the ones about multiple `defaultfor`... have you got any ideas why that would be?